### ISSUE TITLE
Max-plus matrix multiplication

### ISSUE THREAD

**Issue by Jogima-cyber** [getkeops/keops#365]
Hi there! I'm currently looking for an autodiff approach to perform a max-plus matrix multiplication on cuda:
<img width="217" alt="Capture d’écran 2024-03-18 à 12 29 29" src="https://github.com/getkeops/keops/assets/58483181/f521b09e-8bc2-42d6-837f-2a123acab77e">
I've been experimenting with KeOps, but it seems it might not be the ideal tool for this task. Do you think there's a possibility of tweaking it to accommodate this specific operation?

**Comment by jeanfeydy** [comment]
Hi @Jogima-cyber ,

Thanks for your interest in our library.
I agree with you that KeOps does not seem to be a great fit here, except if the number of k-indices is small (< 100) and you want to perform a symbolic computation on your matrix $C_{ij}$. The code below lets you compute the dense `(I, J)` matrix, but is far from being optimal on GPU as it cannot leverage the block structure efficiently:

```python
import torch
from pykeops.torch import LazyTensor

I, J, K = 1000, 2000, 3000
a = torch.randn(I, K)
b = torch.randn(J, K)

a_ = LazyTensor(a.view(I, 1, 1, K, 1))
b_ = LazyTensor(b.view(1, J, 1, K, 1))
maxplus = (a_ + b_).max(dim=3)
print(maxplus.shape)  # (I, J, 1, 1)

# Use argmax + indexing to get a differentiable result
```

If you want to write an optimal CUDA kernel for this task, I would suggest looking at [CUTLASS](https://github.com/NVIDIA/cutlass) and [Triton](https://github.com/openai/triton).

Going forward, we could support a better (but still sub-optimal) implementation by allowing the indexation of KeOps variables by other, integer-valued variables.
It would be a very useful and idiomatic improvement for KeOps, and code could look like:

```python
a_ = LazyTensor(a.view(I, 1, K, 1))
b_ = LazyTensor(b.view(1, J, 1, K))
k_ = LazyTensor(torch.arange(K).view(1, 1, K, 1))
maxplus = (a_ + b_[k_]).max(dim=2)
print(maxplus.shape)  # (I, J, 1)
```

Performance would be OK if K < 100, and we could cut the problem in smaller chunks for larger values of K.
I don't know if @joanglaunes want to comment on this?

Best regards,
Jean

**Comment by joanglaunes** [comment]
Hello,
I have thought about another option, which is to write
```python
a_ = LazyTensor(a.view(I, 1, 1, K))
b_ = LazyTensor(b.view(1, J, 1, K))
maxplus = (a_ + b_).max(dim=3).sum(dim=2)
```
Here the summation along dim 2 does nothing since the size is 1, but it is just a dummy reduction to convert LazyTensor to actual tensor. Trying it, this option appears to be faster, but actually there seems to be an issue with this dummy summation, we need to investigate.

**Comment by Jogima-cyber** [comment]
Hello! @jeanfeydy @joanglaunes thank you very much for the time you spent trying several approaches to solve this issue! I tried @jeanfeydy's solution for my use case, and while it was quite slow, it was sufficient to test an idea. If the idea shows interesting results, I can always write an optimized cuda kernel as suggested by @jeanfeydy. Thank you again for your time.

**Comment by jeanfeydy** [comment]
Hi @Jogima-cyber , you're very welcome! As far as I can tell, [Taichi](https://www.taichi-lang.org/) would be a good tool to implement such an operation efficiently. It is lower-level than KeOps (see e.g. [this example](https://github.com/taichi-dev/taichi_benchmark/blob/main/suites/n_body/taichi/nbody_cache_block.py) that implements the same [tiled reduction scheme](http://kernel-operations.io/keops/engine/map_reduce_schemes.html) as we do), but still much more readable than C++ and easier to deploy. Best of luck :-)

**Comment by joanglaunes** [comment]
Hello @Jogima-cyber , @jeanfeydy ,
I have added the support for the indexing of a `LazyTensor` by another `LazyTensor`, as Jean was suggesting. This was requested also in issue #367 . It is currently in a branch called `Index`. Maybe Jean you can try your idea to see if it works as expected ?
